{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from tqdm import tqdm\n",
    "from staphb_toolkit.lib import calldocker as container_engine\n",
    "from staphb_toolkit.lib.autopath import path_replacer\n",
    "import staphb_toolkit.lib.container_handler as container\n",
    "source_dir = f'data/0_raw'\n",
    "target_dir = f'data/1_trimmed'\n",
    "if not os.path.isdir(target_dir):\n",
    "    os.makedirs(target_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in tqdm(os.listdir(source_dir)):\n",
    "    id = file.split(\"_\")[0]\n",
    "    print(id)\n",
    "    if os.path.exists(f'{target_dir}/{id}_1.trimd.fastq'):\n",
    "        print(f'{id} already trimmed, continuing')\n",
    "        continue\n",
    "    print(f'trimming {id}')\n",
    "    command = f'staphb-tk trimmomatic PE {source_dir}/{id}_1.fastq {source_dir}/{id}_2.fastq {target_dir}/{id}_1.paired.fastq {target_dir}/{id}_1.unpaired.fastq {target_dir}/{id}_2.paired.fastq {target_dir}/{id}_2.unpaired.fastq LEADING:20 TRAILING:20 MINLEN:50'\n",
    "    os.system(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_dir = f'data/1_trimmed'\n",
    "target_dir = f'data/2_qc'\n",
    "if not os.path.isdir(target_dir):\n",
    "    os.makedirs(target_dir)\n",
    "files = [file.replace(\".fastq\",\"\") for file in os.listdir(source_dir) if file.endswith(\".paired.fastq\")]\n",
    "files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in files:\n",
    "    if os.path.exists(f'{target_dir}/{file}_fastqc.html'):\n",
    "        print(f'{file} already qc, continue')\n",
    "        continue\n",
    "    command = f'staphb-tk fastqc {source_dir}/{file}.fastq -o {target_dir}'\n",
    "    os.system(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "command = f'staphb-tk multiqc data/2_qc -o data/3_mqc --force'\n",
    "os.system(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "source_dir = f'data/1_trimmed'\n",
    "target_dir = f'data/4_assembled'\n",
    "if not os.path.isdir(target_dir):\n",
    "    os.makedirs(target_dir)\n",
    "files = [file.replace(\"_1.paired.fastq\",\"\") for file in os.listdir(source_dir) if file.endswith('1.paired.fastq')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in files:\n",
    "    print(file)\n",
    "    if os.path.isdir(f'{target_dir}/{file}'):\n",
    "        print(f'already assembled {file}')\n",
    "        continue\n",
    "    args = [f'-1',f'{source_dir}/{file}_1.paired.fastq',f'-2',f'{source_dir}/{file}_2.paired.fastq']\n",
    "    arg_string,path_map = path_replacer(args,os.getcwd())\n",
    "    print(arg_string,path_map)\n",
    "    command = f'spades.py {arg_string} -o {target_dir}/{file}'\n",
    "    program_object = container.Run(command=command, path=path_map, image='staphb/spades', tag='latest')\n",
    "    program_object.run()\n",
    "    os.system(f'cp {target_dir}/{file}/contigs.fasta {target_dir}/{file}/{file}_contigs.fasta')\n",
    "command = f'staphb-tk quast data/4_assembled/*/contigs.fasta -o data/5_quast -t 1'\n",
    "os.system(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "command = f'blastn -query contigs.fasta -db pathogen_database/ref_prok_rep_genomes -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle\" -out results.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_1_length_392911_cov_30.838820</td>\n",
       "      <td>gi|1109557564|ref|NC_003197.2|</td>\n",
       "      <td>99.302</td>\n",
       "      <td>100987</td>\n",
       "      <td>696</td>\n",
       "      <td>7</td>\n",
       "      <td>146</td>\n",
       "      <td>101124</td>\n",
       "      <td>1325677</td>\n",
       "      <td>1224692</td>\n",
       "      <td>0.0</td>\n",
       "      <td>182600.0</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_1_length_392911_cov_30.838820</td>\n",
       "      <td>gi|1109557564|ref|NC_003197.2|</td>\n",
       "      <td>99.125</td>\n",
       "      <td>80712</td>\n",
       "      <td>674</td>\n",
       "      <td>15</td>\n",
       "      <td>101068</td>\n",
       "      <td>181774</td>\n",
       "      <td>1224197</td>\n",
       "      <td>1143513</td>\n",
       "      <td>0.0</td>\n",
       "      <td>145100.0</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NODE_1_length_392911_cov_30.838820</td>\n",
       "      <td>gi|1109557564|ref|NC_003197.2|</td>\n",
       "      <td>99.186</td>\n",
       "      <td>75208</td>\n",
       "      <td>594</td>\n",
       "      <td>9</td>\n",
       "      <td>221400</td>\n",
       "      <td>296600</td>\n",
       "      <td>1102604</td>\n",
       "      <td>1027408</td>\n",
       "      <td>0.0</td>\n",
       "      <td>135500.0</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NODE_1_length_392911_cov_30.838820</td>\n",
       "      <td>gi|1109557564|ref|NC_003197.2|</td>\n",
       "      <td>99.293</td>\n",
       "      <td>53853</td>\n",
       "      <td>373</td>\n",
       "      <td>7</td>\n",
       "      <td>317605</td>\n",
       "      <td>371451</td>\n",
       "      <td>962586</td>\n",
       "      <td>908736</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97330.0</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NODE_1_length_392911_cov_30.838820</td>\n",
       "      <td>gi|1109557564|ref|NC_003197.2|</td>\n",
       "      <td>99.381</td>\n",
       "      <td>21170</td>\n",
       "      <td>128</td>\n",
       "      <td>3</td>\n",
       "      <td>371383</td>\n",
       "      <td>392551</td>\n",
       "      <td>906624</td>\n",
       "      <td>885457</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38365.0</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   0                               1       2   \\\n",
       "0  NODE_1_length_392911_cov_30.838820  gi|1109557564|ref|NC_003197.2|  99.302   \n",
       "1  NODE_1_length_392911_cov_30.838820  gi|1109557564|ref|NC_003197.2|  99.125   \n",
       "2  NODE_1_length_392911_cov_30.838820  gi|1109557564|ref|NC_003197.2|  99.186   \n",
       "3  NODE_1_length_392911_cov_30.838820  gi|1109557564|ref|NC_003197.2|  99.293   \n",
       "4  NODE_1_length_392911_cov_30.838820  gi|1109557564|ref|NC_003197.2|  99.381   \n",
       "\n",
       "       3    4   5       6       7        8        9    10        11  \\\n",
       "0  100987  696   7     146  101124  1325677  1224692  0.0  182600.0   \n",
       "1   80712  674  15  101068  181774  1224197  1143513  0.0  145100.0   \n",
       "2   75208  594   9  221400  296600  1102604  1027408  0.0  135500.0   \n",
       "3   53853  373   7  317605  371451   962586   908736  0.0   97330.0   \n",
       "4   21170  128   3  371383  392551   906624   885457  0.0   38365.0   \n",
       "\n",
       "                                                  12  \n",
       "0  Salmonella enterica subsp. enterica serovar Ty...  \n",
       "1  Salmonella enterica subsp. enterica serovar Ty...  \n",
       "2  Salmonella enterica subsp. enterica serovar Ty...  \n",
       "3  Salmonella enterica subsp. enterica serovar Ty...  \n",
       "4  Salmonella enterica subsp. enterica serovar Ty...  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv('results.tsv',delimiter='\\t',header=None,nrows=5)\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
